examples/mauna-loa/script.jl

```diff
@@ -0,0 +1,170 @@
+    ),
+    se2 = (;
+        σ = positive(exp(4.0)),
+        ℓ = positive(exp(0.0)),
+    ),
+    rq = (;
+        σ = positive(exp(0.0)),
+        ℓ = positive(exp(0.0)),
+        α = positive(exp(-1.0)),
+    ),
+    se3 = (;
+        σ = positive(exp(-2.0)),
+        ℓ = positive(exp(-2.0)),
+    ),
+    noise_scale = positive(exp(-2.0)),
+)
+#! format: on
+
+# We define a couple of helper functions to simplify the kernel construction:
+
+begin
+    SE(θ) = θ.σ^2 * with_lengthscale(SqExponentialKernel(), θ.ℓ)
+    # PeriodicKernel is broken, see https://github.com/JuliaGaussianProcesses/KernelFunctions.jl/issues/389
+    #Per(θ) = θ.σ^2 * with_lengthscale(PeriodicKernel(; r=[θ.ℓ/2]), θ.p)  # NOTE- discrepancy with GaussianProcesses.jl
+    function Per(θ)
+        return θ.σ^2 * with_lengthscale(SqExponentialKernel(), θ.ℓ) ∘
+               PeriodicTransform(1 / θ.p)
+    end
+    RQ(θ) = θ.σ^2 * with_lengthscale(RationalQuadraticKernel(; α=θ.α), θ.ℓ)
+end
+
+function build_gp_prior(θ)
+    # Kernel is represented as a sum of kernels
+    kernel = SE(θ.se1) + Per(θ.per) * SE(θ.se2) + RQ(θ.rq) + SE(θ.se3)
+    return GP(kernel)
+end
+
+function build_finite_gp(θ)
+    f = build_gp_prior(θ)
+    return f(xtrain, θ.noise_scale^2)
+end
+
+function build_posterior_gp(θ)
+    fx = build_finite_gp(θ)
+    return posterior(fx, ytrain)
+end
+
+# We can now construct the posterior GP.
+# The call to `ParameterHandling.value` is required to replace the constraints (such as `positive`) with concrete numbers:
+
+fpost_init = build_posterior_gp(ParameterHandling.value(θ_init))
+
+# This is what the GP fitted to the data looks like for the initial choice of kernel hyperparameters:
+
+let
+    plotdata()
+    plot!(fpost_init(1900:0.2:2050))
+end
+
+# To improve the fit, we want to maximize the (log) marginal likelihood with respect to the hyperparameters.
+# Optim.jl expects to minimize a loss, so we define it as the negative log marginal likelihood:
+
+function loss(θ)
+    fx = build_finite_gp(θ)
+    # TODO: define logpdf(::PosteriorGP) in AbstractGPs.jl?
+    lml = logpdf(fx, ytrain)
+    return -lml
+end
+
+# In the future, we are planning to provide the following utility function as
+# part of JuliaGaussianProcesses -- for now, we just define it inline.
+# The L-BFGS parameters were chosen because they seem to work well empirically.
+# You could also try with the defaults.
+
+default_optimizer = LBFGS(;
+    alphaguess=Optim.LineSearches.InitialStatic(; scaled=true),
+    linesearch=Optim.LineSearches.BackTracking(),
+)
+
+function optimize_loss(loss, θ_init; optimizer=default_optimizer, maxiter=1_000)
+    options = Optim.Options(; iterations=maxiter, show_trace=true)
+
+    θ_flat_init, unflatten = ParameterHandling.value_flatten(θ_init)
+    loss_packed = loss ∘ unflatten
+
+    # https://julianlsolvers.github.io/Optim.jl/stable/#user/tipsandtricks/#avoid-repeating-computations
+    function fg!(F, G, x)
+        if F != nothing && G != nothing
```

Comment:
One should basically always use `!== nothing` and `=== nothing`:
```suggestion
        if F !== nothing && G !== nothing
```
Here it seems a bit nicer though to just dispatch, i.e., to define
````julia
function fg!(::Nothing, ::Nothing, x)
...
end
function fg!(f, ::Nothing, x)
...
end
function fg!(::Nothing, G, x)
...
end
```

Reply:
Tell that to the people of Optim.jl! I just copied it from https://julianlsolvers.github.io/Optim.jl/stable/#user/tipsandtricks/#avoid-repeating-computations ...

Reply:
Unrelated from style issues, one should basically never compare `nothing` with `==` or `!=`, so let's change this at least.

Reply:
Also opened a PR upstream: https://github.com/JuliaNLSolvers/Optim.jl/pull/961